WordPressAuthenticator/OAuth/ProofKeyForCodeExchange.swift

```diff
@@ -25,21 +25,72 @@ struct ProofKeyForCodeExchange {
         }
     }
 
-    let codeVerifier: String
+    let codeVerifier: CodeVerifier
     let method: Method
 
-    init(codeVerifier: String, method: Method) {
+    init(codeVerifier: CodeVerifier = CodeVerifier(), method: Method = .s256) {
         self.codeVerifier = codeVerifier
         self.method = method
     }
 
     var codeCallenge: String {
         switch method {
         case .s256:
-            // TODO: code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
-            fatalError()
+            // The spec defines code_challenge for the s256 mode as:
+            //
+            // code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
+            //
+            // We don't need the ASCII conversion, because we build `CodeVerifier` from URL safe
+            // characters.
+            //
+            // In the same way, it is safe to force unwrap the `Data` conversion because the
+            // `CodeVerifier` input is guaranteed to have only UTF-8 representable characters.
+            //
+            // Also worth noting that `.data(using: .utf8)` cannot technically return `nil` anyway.
+            // See https://forums.swift.org/t/can-encoding-string-to-data-with-utf8-fail/22437/4.
+            let rawData = codeVerifier.rawValue.data(using: .utf8)!
+            let hashedData: Data = rawData.sha256Hashed()
+            return hashedData.base64URLEncodedString()
         case .plain:
-            return codeVerifier
+            return codeVerifier.rawValue
         }
     }
 }
+
+extension ProofKeyForCodeExchange {
+
+    struct CodeVerifier: Equatable {
+
+        let rawValue: String
+
+        // From the docs: using the unreserved characters [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~"
+        // That is, URL safe characters.
+        //
+        // Notice that Swift offers `CharacterSet.urlQueryAllowed` to represent this set of characters.
+        // However, there is no straightforward way to convert a `CharacterSet` to a `Set<Character>`.
+        // See for example https://nshipster.com/characterset/.
+        private let allowedCharacters = Set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~")
+        private lazy var allowedCharactersCount = UInt32(allowedCharacters.count)
+
+        /// `length` must be between 43 and 128, inclusive.
+        init(length: Int = 128) {
+            let constrainedLength = min(max(length, 43), 128)
+            rawValue = String.randomString(using: allowedCharacters, withLength: constrainedLength)
+        }
+    }
+}
+
+// This is a helper for the tests.
+//
+// Unfortunately, it needs to be part of the production code because Swift doesn't allow adding
+// non-convenience initializers outside the module.
+extension ProofKeyForCodeExchange.CodeVerifier {
+
+    init?(value: String) {
```

Comment:
I wonder if it'd be worth putting this in the body of the type, and _possibly_ as the default initializer? Then your "make new PKCE" method would be like `ProofKeyForCodeExchange.create(withLength:)`? 

The reason I ask is that do we know for sure that we'll never need to parse an existing one? I've found that an initializer than only ever creates a new value but can't process an existing one eventually gets me in trouble 😅

Reply:
> I wonder if it'd be worth putting this in the body of the type,

See 0361bb24

Reply:
> I've found that an initializer than only ever creates a new value but can't process an existing one eventually gets me in trouble 😅

Ah! Thank you fort the suggestion. I don't think I wrote it but I didn't like the state of initialization either, but at the time I decided to ship as it was rather than spending extra time on it.

You suggestion was just what I needed to come up with the mix of idea and motive to improve the implementation.

See ff9f23d5